### ISSUE TITLE
Feature requests: Client Caching 

### ISSUE THREAD

**Issue by Farshidgolkarihagh** [RSuter/NSwag#861]
It will be super cool to have a client caching which is created based on some sort of attribute in the web api endpoint. For example, let's say you have a typescript client with certain method and arguments. If we can automatically have a caching layer in typescript which caches result based on the parameters used that will be very cool. I know it will be quite complicated as you need support for different client languages.

**Comment by grovesNL** [comment]
It would be interesting, although caching requirements tend to vary between applications though. For that reason I think should be handled through the application rather than NSwag.

One idea could be to use a [`Proxy`](http://2ality.com/2015/10/intercepting-method-calls.html) to wrap the generated NSwag clients in a cache function. As each method is called, you could check whether cached entries already exist for that combination of method/arguments (depending on the type of request, cache policy, etc.).

**Comment by RSuter** [comment]
I agree with josh her: Caching is very use case specific and I think should not be baked into the generated code. However we need some extension points (eg a single RequestAsync method on the base class) where you can implement your caching functionality for all operations. Maybe a new setting CallRequestWrapperMethod which always calls this method. Somthing like: 

    public Task<T> RequestAsync<T>(string operation, Action<Task<T>> execute)

The action execute calls the original/generated implementation...

What do you think?

Ref: https://github.com/RSuter/NSwag/issues/321

**Comment by Farshidgolkarihagh** [comment]
Hi

Yeah I agree with Josh. That would simplify the integration and
customisation a lot. Not sure what is the operation parameters!? So how
would you add caching for some method and not others.!? it will be good to
not have to write various IF statement to only cache certain API calls. I
think it will be good to have a attribute on the endpoint method to
indicate the endpoint to cache but how it is cached is up to the developer.


What do you think?

Kind regard,
Farshid

**Comment by Worthy7** [comment]
Hi 2019 here, need this!

**Comment by thtp** [comment]
May be it would be enough to just add an option to make client methods virtual. With virtual methods user will be able to inherit client class and wrap base calls for any purpose, including caching.

**Comment by frank-hoffman** [comment]
> May be it would be enough to just add an option to make client methods virtual. With virtual methods user will be able to inherit client class and wrap base calls for any purpose, including caching.

This was a good idea in 2020 and is still a good idea in 2023. The inability to incorporate caching is what's holding me back from using NSwag right now.
